uproot/source/chunked.py

```diff
@@ -106,21 +108,29 @@ def data(self, start, stop, dtype=None):
                 future = self._futures.pop(chunkindex, None)
                 if future is not None:
                     chunk = future.result()
-                    if chunk is not None:
-                        self.cache[chunkindex] = chunk
 
             if chunk is None:
                 try:
                     chunk = self.cache[chunkindex]
                 except KeyError:
                     self._open()
-                    chunk = self.cache[chunkindex] = self._read(chunkindex)
+                    chunk = self._read(chunkindex)
 
             cstart = 0
             cstop = self._chunkbytes
             gstart = chunkindex * self._chunkbytes
             gstop = (chunkindex + 1) * self._chunkbytes
 
+            if len(chunk) > self._chunkbytes:
+                warnings.warn('Received larger chunk than expected, assumed to be entire file. Performance may be significantly degraded.', RuntimeWarning)
```

Comment:
I'd rather not have warnings, even if performance is significantly degraded. (There's an uncountable number of ways to do _that_, after all.) Exceptions in Python are fine—they tell you where to go, or at least start looking, to fix them—but warnings are "sticky"—you see them but don't know how to fix them (if you don't search the codebase for the text of the warning).

Reply:
The issue with performance here when the cache size limit is smaller than the entire file. My fix is very naive and will just keep downloading the same file over and over and over again while uproot seeks through it.

Now I've thought about it again, I've just pushed an improvement to make the situation slightly less bad. I also now ensure that even if the current chunk is evicted from the cache it will still work.

I see three options for how to handle this case:
1. Only show the warning if the file is larger than the cache? To me it feels like enough of a footgun that users should know but you're right that it isn't clear what to do. I've just done this and edited the message to be more explicit.
2. Alternatively as the entire file is already in memory at this point, the cache could be replaced with a simple dictionary containing the entire file to waste memory instead of time?
3. Drop the warning entirely like you first suggested.

Which would you prefer?

Reply:
See comments below.